compiler/rustc_borrowck/src/lib.rs

```diff
@@ -1100,13 +1100,22 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                                 this.report_conflicting_borrow(location, place_span, bk, borrow);
                             this.buffer_error(err);
                         }
-                        WriteKind::StorageDeadOrDrop => this
-                            .report_borrowed_value_does_not_live_long_enough(
-                                location,
-                                borrow,
-                                place_span,
-                                Some(kind),
-                            ),
+                        WriteKind::StorageDeadOrDrop => {
+                            if let Some(DesugaringKind::Replace) = place_span.1.desugaring_kind() {
```

Comment:
This may deserve a helper function. The fact that checking `desugaring_kind()` is enough comes from the fact that `DesugaringKind::Replace` can only be created *after the other (HIR-based) variants*.